src/core_ocean/shared/mpas_ocn_vmix_gotm.F

```diff
@@ -0,0 +1,494 @@
+         waterDepth = 0.0_RKIND
+         do k = minLevelCell(iCell), maxLevelCell(iCell)
+            NSquaredTop(k) = BruntVaisalaFreqTop(k,iCell)
+            hLayer(k) = layerThickness(k,iCell)
+            waterDepth = waterDepth + layerThickness(k,iCell)
+         end do
+
+         ! compute the bottom friction velocity
+         bottomFrictionVelocity = sqrt(2.0_RKIND * gotmBottomDragCoef *  &
+                                       kineticEnergyCell(maxLevelCell(iCell),iCell))
+
+         ! fill GOTM arrays with current values at iCell
+         do k = 0, maxLevelCell(iCell) - minLevelCell(iCell) + 1
+            ki = maxLevelCell(iCell) - k + 1
+            gotm_vertVisc(k) = gotmVertViscTopOfCellCur(ki, iCell)
+            gotm_vertDiffT(k) = gotmVertDiffTopOfCellCur(ki, iCell)
+            gotm_vertDiffS(k) = gotmVertDiffTopOfCellCur(ki, iCell)
+            gotm_TKE(k) = gotmTKETopOfCellCur(ki, iCell)
+            gotm_eps(k) = gotmDissTopOfCellCur(ki, iCell)
+            gotm_L(k) = gotmLengthTopOfCellCur(ki, iCell)
+            gotm_kb(k) = gotmKbTopOfCellCur(ki, iCell)
+            gotm_epsb(k) = gotmEpsbTopOfCellCur(ki, iCell)
+         end do
+
+         !
+         ! interface with do_turbulence():
+         !
+         !   subroutine do_turbulence(nlev,dt,depth,u_taus,u_taub,z0s,z0b,h,NN,SS,xP)
+         !
+         !   nlev          : number of vertical layers
+         !   dt            : time step (s)
+         !   depth         : water depth (m)
+         !   u_taus        : surface friction velocity (m/s)
+         !   u_taub        : bottom friction velocity (m/s)
+         !   z0s           : surface roughness length (m)
+         !   z0b           : bottom roughness length (m)
+         !   h             : layer thickness (m)
+         !   NN            : bouyancy frequency squared (1/s^2)
+         !   SS            : shear-frequency squared (1/s^2)
+         !   xP (optional) : extra TKE production due to, e.g., seagrass (m^2/s^3)
+         !
+         ! MPAS-O (1:nlev+1) <-> GOTM (nlev:0)
+         !
+         call do_turbulence(maxLevelCell(iCell) - minLevelCell(iCell) + 1,                &
+                            gotm_dt,                                                      &
+                            waterDepth,                                                   &
+                            surfaceFrictionVelocity(iCell),                               &
+                            bottomFrictionVelocity,                                       &
+                            gotmSurfaceRoughnessLength,                                   &
+                            gotmBottomRoughnessLength,                                    &
+                            hLayer(maxLevelCell(iCell)+1:minLevelCell(iCell):-1),         &
+                            NSquaredTop(maxLevelCell(iCell)+1:minLevelCell(iCell):-1),    &
+                            shearSquaredTop(maxLevelCell(iCell)+1:minLevelCell(iCell):-1))
```

Comment:
@qingli411 when the order is inverted here, the compiler is forced to create temporary arrays in memory within the `iCell` loop. This is bad for performance and unnecessary, since these three arrays are created above. Please change these three variables to be ordered from bottom to top above, and then pass them in without the `-1` index at the end. You can test it by compiling with intel debug. Temporary array creation produces a warning like this:
```
forrtl: warning (406): fort: (1): In call to DO_TURBULENCE, an array temporary was created for argument #10

Image              PC                Routine            Line        Source
ocean_model        0000000002E0D2C6  Unknown               Unknown  Unknown
ocean_model        00000000016EDAD2  ocn_vmix_gotm_mp_         286  mpas_ocn_vmix_gotm.f90
ocean_model        00000000016F701F  ocn_vmix_mp_ocn_v         188  mpas_ocn_vmix.f90
```
or you can just let me test again.

Reply:
@mark-petersen I redefined these arrays with indices in the GOTM order, which starts from 0. Could you please test it again and see if there are any problems?